OutOfSchool/OutOfSchool.WebApi/Services/AchievementService.cs

```diff
@@ -0,0 +1,152 @@
+
+            foreach (string Teacher in dto.Teachers)
+            {
+                achievement.Teachers.Add(new AchievementTeacher { Title = Teacher, Achievement = achievement });
+            }
+
+            var newAchievement = await repository.Create(achievement).ConfigureAwait(false);
+
+            logger.LogInformation($"Achievement with Id = {newAchievement?.Id} created successfully.");
+
+            return mapper.Map<AchievementDto>(newAchievement);
+        }
+
+        /// <inheritdoc/>
+        public async Task<Result<AchievementDto>> Delete(Guid id)
+        {
+            logger.LogInformation($"Deleting Achievement with Id = {id} started.");
+
+            var entity = new Achievement() { Id = id };
+
+            try
+            {
+                await repository.Delete(entity).ConfigureAwait(false);
+
+                logger.LogInformation($"Achievement with Id = {id} succesfully deleted.");
+
+                return Result<AchievementDto>.Success(mapper.Map<AchievementDto>(entity));
+            }
+            catch (DbUpdateConcurrencyException)
+            {
+                logger.LogError($"Deleting failed. Achievement with Id = {id} doesn't exist in the system.");
+                throw;
+            }
+        }
+
+        /// <inheritdoc/>
+        public async Task<AchievementDto> GetById(Guid id)
+        {
+            logger.LogInformation($"Getting Achievement by Id started. Looking Id = {id}.");
+
+            var achievement = await repository.GetById(id).ConfigureAwait(false);
+
+            if (achievement == null)
+            {
+                throw new ArgumentOutOfRangeException(
+                    nameof(id),
+                    localizer["The id cannot be greater than number of table entities."]);
+            }
+
+            logger.LogInformation($"Successfully got a Achievement with Id = {id}.");
+
+            return mapper.Map<AchievementDto>(achievement);
+        }
+
+        /// <inheritdoc/>
+        public async Task<AchievementDto> Update(AchievementCreateDTO dto)
+        {
+            logger.LogInformation($"Updating Achievement with Id = {dto?.Id} started.");
+
+            try
+            {
+                var achievement = await repository.GetById(dto.Id).ConfigureAwait(false);
+
+                achievement.AchievementTypeId = dto.AchievementTypeId;
+                achievement.Title = dto.Title;
+                achievement.AchievementDate = dto.AchievementDate;
+                achievement.WorkshopId = dto.WorkshopId;
+                achievement.Children.RemoveAll(x => dto.ChildrenIDs.IndexOf(x.Id) < 0);
+                achievement.Children.AddRange(context.Children.Where(w => dto.ChildrenIDs.Contains(w.Id)).ToList());
+
+                achievement.Teachers.RemoveAll(x => dto.Teachers.IndexOf(x.Title) < 0);
+                foreach (var Teacher in dto.Teachers)
+                {
+                    if (achievement.Teachers.Find(x => x.Title == Teacher) is null)
```

Comment:
You could rewrite something like this.
```
var exceptTeachers = dto.Teachers
   .Where(p => achievement.Teachers.All(x => !x.Title.Equals(p)));
foreach (var teacherId in exceptTeachers)
{
   achievement.Teachers.Add(new AchievementTeacher { Title = dto.Title, Achievement = achievement });
}
```
**Should compare strings by `Equals()` method**